bpo-30028: make test.support.temp_cwd() fork-safe

akruis opened the pull request:
Improve the context manager test.support.temp_cwd() to not remove the temporary
directory, if a forked child terminates.

<!-- issue-number: bpo-30028 -->
https://bugs.python.org/issue30028
<!-- /issue-number -->

Commit:
bpo-30028: make test.support.temp_cwd() fork-safe

Improve the context manager test.support.temp_cwd() to not remove the temporary
directory, if a forked child terminates.

serhiy-storchaka commented on Lib/test/test_regrtest.py:
```diff
@@ -821,5 +821,20 @@ def test_crashed(self):
+class TempCwdTestCase(unittest.TestCase):
+    @unittest.skipUnless(hasattr(os, "fork"), "test requires os.fork")
+    def test_forked_child(self):
+        import sys
```
Is it used?

akruis replied:
Opps, a left over from a debug print to sys.stderr

Commit:
Update test_regrtest.py

Remove an unused "import sys".

vstinner commented on Lib/test/support/__init__.py:
```diff
@@ -960,10 +960,12 @@ def temp_dir(path=None, quiet=False):
         yield path
     finally:
-        if dir_created:
+        if dir_created and pid == os.getpid():
```
The test is non obvious, can you please add a short comment explaining why you test the pid?

I guess that it's test to avoid that two processes try to remove the directory if the parent uses fork()?

Add bpo-30028 in the comment.

vstinner commented on Lib/test/test_regrtest.py:
```diff
@@ -821,5 +821,19 @@ def test_crashed(self):
                                   randomize=True)
 
 
+class TempCwdTestCase(unittest.TestCase):
```
Please move this test to test_support.py which tests test.support.

test_regrtest tests Lib/test/libregrtest/.

Commit:
bpo-30028: Add comments, move the test to test_support.py

Added an explanatory comment and moved the test to test_support.py.
I also modified the test to be in line with the other tests of temp_cwd.

vstinner commented on Lib/test/test_support.py:
```diff
@@ -161,6 +161,20 @@ def test_temp_dir__existing_dir__quiet_true(self):
+    @unittest.skipUnless(hasattr(os, "fork"), "test requires os.fork")
+    def test_temp_dir__forked_child(self):
+        """Test that a forked child process does not remove the directory."""
+        with support.temp_cwd() as temp_path:
```
Hum, please don't use fork in a test process. Write a short script run with test.support.script_helper. Otherwise, I expect too many issues like bad handling of KeyboardInterrupt. Use textwrap.dedent() to include the script directly here.

akruis replied:
Done (1a0364d4). Is the patch now OK?

Commit:
bpo-30028: Run the fork test as an external script.

vstinner commented on Lib/test/test_support.py:
```diff
@@ -161,6 +163,22 @@ def test_temp_dir__existing_dir__quiet_true(self):
+                    # parent process
+                    os.waitpid(pid, 0)  # wait for the child to terminate
+                    # make sure that temp_path is still present
+                    assert os.path.isdir(temp_path), "Child removed temp_path."
```
I would avoid assert keyword which might be removed depending on the Python option, but use an explicit if + raise Exception(...). But maybe it's ok, I don't know :-)

serhiy-storchaka replied:
Since this is a subprocess, explicit `os.exit()` can be more appropriate.

akruis replied:
I'm currently on a business trip and can't access my Linux box. I'll update the pull request on Thursday. But could you please agree on either raising AssertionError or a print() and explicit os.exit(). In the end it makes no difference.

Commit:
bpo-30028: replace assert by if ...: raise AssertionError

An assert might removed depending on Python options.

akruis commented:
@haypo and @serhiy-storchaka: Any chance to get this PR merged?

vstinner commented on Lib/test/test_support.py:
```diff
@@ -161,6 +163,23 @@ def test_temp_dir__existing_dir__quiet_true(self):
+                pid = os.fork()
+                if pid != 0:
+                    # parent process
+                    os.waitpid(pid, 0)  # wait for the child to terminate
```
Please test that the child didn't fail.

vstinner commented on Lib/test/test_support.py:
```diff
@@ -161,6 +163,23 @@ def test_temp_dir__existing_dir__quiet_true(self):
+                    # make sure that temp_path is still present
+                    if not os.path.isdir(temp_path):
+                        raise AssertionError("Child removed temp_path.")
+        """))
```
Please add a comment explaining that the child exits without removing the temporary directory.

serhiy-storchaka replied:
Isn't the comment above enough?

vstinner commented on Lib/test/test_support.py:
```diff
@@ -161,6 +163,23 @@ def test_temp_dir__existing_dir__quiet_true(self):
 
+    @unittest.skipUnless(hasattr(os, "fork"), "test requires os.fork")
+    def test_temp_dir__forked_child(self):
+        """Test that a forked child process does not remove the directory."""
```
Please add a reference to the issue: bpo-30028

akruis replied:
No other doc string in this file contains a reference to bpo. Therefore I added the reference as a comment.

Commit:
bpo-30028: improve the test case

- better comments
- make sure, that the child exits successfully

serhiy-storchaka committed:
Merge branch 'master' into tempcwd_fork_fix

akruis committed:
bpo-30028: Add name to Misc/ACKS

Proposed by Serhiy Storchaka.
